drivers/i2c/i2c_rcar.c

```diff
@@ -0,0 +1,353 @@
+		k_busy_wait(USEC_PER_MSEC);
+		timeout++;
+	}
+	if (timeout == 2) {
+		return -ETIMEDOUT;
+	}
+
+	i2c_rcar_write(config, RCAR_I2C_ICMAR, (chip << 1) | read);
+	/* Reset */
+	i2c_rcar_write(config, RCAR_I2C_ICMCR, RCAR_I2C_ICMCR_MDBS | RCAR_I2C_ICMCR_MIE | RCAR_I2C_ICMCR_ESG);
+	/* Clear Status */
+	i2c_rcar_write(config, RCAR_I2C_ICMSR, 0);
+
+	timeout = 0;
+	while (!(i2c_rcar_read(config, RCAR_I2C_ICMSR) & mask) && (timeout<100)){
+		k_busy_wait(USEC_PER_MSEC);
+		timeout++;
+	}
+	if (timeout == 100) {
+		return -ETIMEDOUT;
+	}
+
+	/* Check NAK */
+	if(i2c_rcar_read(config, RCAR_I2C_ICMSR) & RCAR_I2C_ICMSR_MNR)
+		return -EIO;
+
+	return 0;
+}
+
+static int i2c_rcar_transfer_msg(const struct device *dev, struct i2c_msg *msg){
+	const struct i2c_rcar_cfg *config = DEV_I2C_CFG(dev);
+	uint32_t icmcr = RCAR_I2C_ICMCR_MDBS | RCAR_I2C_ICMCR_MIE;
+	uint32_t i;
+	uint16_t timeout = 0;
+
+	if ((msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ){
+		for (i = 0; i < msg->len; i++) {
+			if (msg->len - 1 == i) {
+				icmcr |= RCAR_I2C_ICMCR_FSB;
+			}
+
+			i2c_rcar_write(config, RCAR_I2C_ICMCR, icmcr);
+			i2c_rcar_write(config, RCAR_I2C_ICMSR, (uint32_t)~RCAR_I2C_ICMSR_MDR);
+
+			/* Wait for a data to be received*/
+			while (!(i2c_rcar_read(config, RCAR_I2C_ICMSR) & RCAR_I2C_ICMSR_MDR) && (timeout<100)){
+				k_busy_wait(USEC_PER_MSEC);
+				timeout++;
+			}
+			if (timeout == 100) {
+				return -ETIMEDOUT;
+			}
+
+			msg->buf[i] = i2c_rcar_read(config, RCAR_I2C_ICRXD_ICTXD) & 0xff;
+		}
+
+		i2c_rcar_write(config, RCAR_I2C_ICMSR, (uint32_t)~RCAR_I2C_ICMSR_MDR);
+	}
+	else {
+		for (i = 0; i < msg->len; i++) {
+			i2c_rcar_write(config, RCAR_I2C_ICRXD_ICTXD, msg->buf[i]);
+			i2c_rcar_write(config, RCAR_I2C_ICMCR, icmcr);
+			i2c_rcar_write(config, RCAR_I2C_ICMSR, (uint32_t)~RCAR_I2C_ICMSR_MDE);
+
+			while (!(i2c_rcar_read(config, RCAR_I2C_ICMSR) & RCAR_I2C_ICMSR_MDE) && (timeout<100)){
+				k_busy_wait(USEC_PER_MSEC);
+				timeout++;
+			}
+			if (timeout == 100) {
+				return -ETIMEDOUT;
+			}
+		}
+
+		i2c_rcar_write(config, RCAR_I2C_ICMSR, (uint32_t)~RCAR_I2C_ICMSR_MDE);
+		icmcr |= RCAR_I2C_ICMCR_FSB;
+		i2c_rcar_write(config, RCAR_I2C_ICMCR, icmcr);
+	}
+
+	return i2c_rcar_finish(dev);
+}
+
+static int i2c_rcar_transfer(const struct device *dev,
+			   struct i2c_msg *msgs, uint8_t num_msgs,
+			   uint16_t addr)
+{
+	const struct i2c_rcar_cfg *config = DEV_I2C_CFG(dev);
+	uint16_t timeout = 0;
+
+	if (!num_msgs) {
+		return 0;
+	}
+
+	/* Wait for the bus to be available */
+	while ((i2c_rcar_read(config, RCAR_I2C_ICMCR) & RCAR_I2C_ICMCR_FSDA) && (timeout<10)){
+		k_busy_wait(USEC_PER_MSEC);
+		timeout++;
+	}
+	if (timeout == 10) {
+		return -EIO;
+	}
+
+	do {
+		/* Send address after any Start condition */
+		if (i2c_rcar_set_addr(dev, addr, !!(msgs->flags & I2C_MSG_READ))) return -EIO; /* No ACK received */
+
+		/* Transfer data */
+		if (msgs->len) {
+			i2c_rcar_transfer_msg(dev, msgs);
```

Comment:
you do not check return of this function, we may stop sending msg if one failed ?